providers/supermicro/supermicro.go

```diff
@@ -0,0 +1,343 @@
+package supermicro
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/bmc-toolbox/bmclib/v2/internal/httpclient"
+	"github.com/bmc-toolbox/bmclib/v2/providers"
+	"github.com/go-logr/logr"
+	"github.com/jacobweinstock/registrar"
+	"github.com/pkg/errors"
+
+	bmclibconsts "github.com/bmc-toolbox/bmclib/v2/constants"
+	bmclibErrs "github.com/bmc-toolbox/bmclib/v2/errors"
+)
+
+const (
+	// ProviderName for the provider Supermicro implementation
+	ProviderName = "supermicro"
+	// ProviderProtocol for the provider supermicro implementation
+	ProviderProtocol = "vendorapi"
+)
+
+var (
+	// Features implemented
+	Features = registrar.Features{
+		providers.FeatureScreenshot,
+	}
+)
+
+// supports
+// SYS-5019C-MR
+// SYS-510T-MR
+// 6029P-E1CR12L
+
+type Config struct {
+	HttpClient           *http.Client
+	Port                 string
+	httpClientSetupFuncs []func(*http.Client)
+}
+
+// Option for setting optional Client values
+type Option func(*Config)
+
+func WithHttpClient(httpClient *http.Client) Option {
+	return func(c *Config) {
+		c.HttpClient = httpClient
+	}
+}
+
+// WithSecureTLS returns an option that enables secure TLS with an optional cert pool.
+func WithSecureTLS(rootCAs *x509.CertPool) Option {
+	return func(c *Config) {
+		c.httpClientSetupFuncs = append(c.httpClientSetupFuncs, httpclient.SecureTLSOption(rootCAs))
+	}
+}
+
+func WithPort(port string) Option {
+	return func(c *Config) {
+		c.Port = port
+	}
+}
+
+// Connection details
+type Client struct {
+	client    *http.Client
+	host      string
+	user      string
+	pass      string
+	port      string
+	csrfToken string
+	log       logr.Logger
+}
+
+// New returns connection with a Supermicro client initialized
+func NewClient(host, user, pass string, log logr.Logger, opts ...Option) *Client {
```

Comment:
might think about making `log` optional a then default to `logr.Discard`. Not a PR blocker.